File: librarymanagement/src/main/java/com/example/librarymanagement/service/BookService.java
package com.example.librarymanagement.service;


import com.example.librarymanagement.dto.BookDto;
import com.example.librarymanagement.dto.BookRequest;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;

public interface BookService {

    Page<BookDto> getAllBooks(Pageable pageable);

    BookDto getBookById(Long id);

    List<BookDto> searchBooks(String title, String author, String isbn, String category);

    BookDto createBook(BookRequest request);

    BookDto updateBook(Long id, BookRequest request);

    void deleteBook(Long id);
}

File: frontend/src/app/books/services/book.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Book, BookRequest } from '../models/book';
import { environment } from '../../../environments/environment';

@Injectable({
  providedIn: 'root'
})
export class BookService {

  private apiUrl = `${environment.apiUrl}/books`;

  constructor(private http: HttpClient) {}

  getAllBooks(): Observable<Book[]> {
    return this.http.get<Book[]>(this.apiUrl);
  }

  getBookById(id: number): Observable<Book> {
    return this.http.get<Book>(`${this.apiUrl}/${id}`);
  }

  searchBooks(title?: string, author?: string,
              isbn?: string, category?: string): Observable<Book[]> {
    let params = new HttpParams();
    if (title)    params = params.set('title', title);
    if (author)   params = params.set('author', author);
    if (isbn)     params = params.set('isbn', isbn);
    if (category) params = params.set('category', category);
    return this.http.get<Book[]>(`${this.apiUrl}/search`, { params });
  }

  createBook(book: BookRequest): Observable<Book> {
    return this.http.post<Book>(this.apiUrl, book);
  }

  updateBook(id: number, book: BookRequest): Observable<Book> {
    return this.http.put<Book>(`${this.apiUrl}/${id}`, book);
  }

  deleteBook(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}